### PR TITLE
Improved keyboard handling and view positioning for the iPhone X / iOS 11.

### PR DESCRIPTION
**Pull request by Elland**
Long live the notch. 👀

### CODE REVIEW

File: Toshi/Controllers/Browse/BrowseViewController.swift [hunk 1]
```diff
@@ -154,7 +180,6 @@ class BrowseViewController: SearchableCollectionController {
         super.viewWillDisappear(animated)
         
         searchBar.resignFirstResponder()
-        clearSearch()
```
Comment (Elland) [inline Toshi/Controllers/Browse/BrowseViewController.swift]: Slipped in this big quality of life improvement.

File: Toshi/Controllers/Chat/ChatViewController.swift [hunk 2]
```diff
@@ -242,11 +253,8 @@ final class ChatViewController: UIViewController, UINavigationControllerDelegate
     }
 
     private func updateConstraints() {
-        textInputViewBottomConstraint?.constant = heightOfKeyboard < -textInputHeight ? heightOfKeyboard + textInputHeight + ChatButtonsView.height : 0
+        textInputViewBottomConstraint?.constant = keyboardScreenPositionOffset // < -textInputHeight ? heightOfKeyboard + textInputHeight + ChatButtonsView.height : 0
```
Comment (designatednerd) [inline Toshi/Controllers/Chat/ChatViewController.swift]: does the stuff commented out here and at line 240 still need to be here?

File: Toshi/Controllers/Browse/BrowseViewController.swift (hunk 1 above)
Comment (designatednerd) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: how does this improve quality of life? 😄

File: Toshi/Controllers/Browse/BrowseViewController.swift [hunk 3]
```diff
@@ -117,6 +117,32 @@ class BrowseViewController: SearchableCollectionController {
...
+        }
+
+        NotificationCenter.default.addObserver(forName: .UIKeyboardWillHide, object: nil, queue: .main) { [weak self] notification in
+            guard let info = notification.userInfo else { return }
+            let duration = info[UIKeyboardAnimationDurationUserInfoKey] as? TimeInterval ?? 0.2
+            let curveRaw = info[UIKeyboardAnimationCurveUserInfoKey] as? UInt ?? 0
+            let curve = UIViewAnimationOptions(rawValue: curveRaw)
+
+            UIView.animate(withDuration: duration, delay: 0, options: UIViewAnimationOptions(rawValue: UInt(curve.rawValue)), animations: {
+                self?.searchResultView.contentInset.bottom = 60
```
Comment (designatednerd) [inline Toshi/Controllers/Browse/BrowseViewController.swift]: where is `60` coming from?

File: Toshi/Controllers/Browse/BrowseViewController.swift (hunk 1 above)
Comment (Elland) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: You don't lose your search term and results every time you leave the screen.

File: Toshi/Controllers/Chat/ChatViewController.swift (hunk 2 above)
Comment (Elland) [reply Toshi/Controllers/Chat/ChatViewController.swift]: Gone!

File: Toshi/Controllers/Chat/ChatViewController.swift [hunk 4]
```diff
@@ -124,9 +124,9 @@ final class ChatViewController: UIViewController, UINavigationControllerDelegate
         let activeNetworkViewHeight = activeNetworkView.heightConstraint?.constant ?? 0
         let topInset = ChatFloatingHeaderView.height + 64.0 + activeNetworkViewHeight
         
-        // The table view is inverted 180 degrees
-        tableView.contentInset = UIEdgeInsets(top: 10, left: 0, bottom: topInset + 2 + 10, right: 0)
-        tableView.scrollIndicatorInsets = UIEdgeInsets(top: 0, left: 0, bottom: topInset + 2, right: 0)
+        // The tableview is inverted 180 degrees
```
Comment (designatednerd) [inline Toshi/Controllers/Chat/ChatViewController.swift]: nitpick: when used in prose it should be two words. If you want it to be one word it should be `UITableView`

File: Toshi/Controllers/Chat/ChatViewController.swift [hunk 5]
```diff
@@ -124,9 +124,9 @@ final class ChatViewController: UIViewController, UINavigationControllerDelegate
...
+        tableView.contentInset = UIEdgeInsets(top: buttonsView.frame.height + 10, left: 0, bottom: topInset + 2 + 10, right: 0)
```
Comment (designatednerd) [inline Toshi/Controllers/Chat/ChatViewController.swift]: anywhere the hard-coded numbers here can be pulled into variables that explain what the hell they're doing?

Comment (Elland) [reply Toshi/Controllers/Chat/ChatViewController.swift]: I don't know where they were coming before, I just made it a bit less horrible and fit. :(

Comment (designatednerd) [reply Toshi/Controllers/Chat/ChatViewController.swift]: k, will let this slide for now

**Comment by designatednerd** [comment]
I thought I saw code somewhere in Sweet UIKit that was REAL similar to this - what's the difference with the keyboard handling changes here?

File: Toshi/Controllers/Browse/BrowseViewController.swift (hunk 3 above)
Comment (Elland) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: It shouldn't be there anymore. Gone. 🚀

**Comment by Elland** [comment]
This uses `SweetUIKit`'s keyboard handling system. 👍

**Comment by designatednerd** [comment]
@Elland Wait then why does the Browse VC still have all the setup for notification listeners?

**Comment by Elland** [comment]
We're only using the live-tracking/interactive dismissing keyboard for chat. Browse only dismisses if you cancel or hit `search`. Not sure if it's a bug or by design, so I didn't change.

**Comment by Elland** [comment]
(The `SweetUIKit` system is only used to have a keyboard-attached view).

**Pull request merged**